File: mycollab-ui/src/main/java/com/esofthead/mycollab/vaadin/resources/file/VaadinFileResource.java
/**
 * This file is part of mycollab-ui.
 *
 * mycollab-ui is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-ui is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-ui.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.vaadin.resources.file;

import com.esofthead.mycollab.configuration.FileStorage;
import com.esofthead.mycollab.vaadin.resources.VaadinResource;
import com.vaadin.server.DownloadStream;
import com.vaadin.server.FileResource;
import com.vaadin.server.Resource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;

/**
 * @author MyCollab Ltd.
 * @since 4.5.1
 */
public class VaadinFileResource extends VaadinResource {

    private static final Logger LOG = LoggerFactory.getLogger(VaadinFileResource.class);

    @Override
    public Resource getStreamResource(String documentPath) {
        return new FileStreamDownloadResource(documentPath);
    }

    static class FileStreamDownloadResource extends FileResource {
        private static final long serialVersionUID = 1L;

        FileStreamDownloadResource(String documentPath) {
            super(new File(FileStorage.baseContentFolder, documentPath));
        }

        @Override
        public DownloadStream getStream() {
            final String fileName = getFilename().replaceAll(" ", "_").replaceAll("-", "_");
            try {
                FileInputStream inStream = new FileInputStream(getSourceFile());
                DownloadStream ds = new DownloadStream(inStream, getMIMEType(), fileName);
                ds.setParameter("Content-Disposition", "attachment; filename=" + fileName);
                ds.setCacheTime(0);
                return ds;
            } catch (IOException e) {
                LOG.error("Error to create download stream", e);
                return null;
            }
        }

    }
}

File: mycollab-web/src/main/java/com/mycollab/vaadin/resources/file/VaadinFileResource.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.vaadin.resources.file

import com.mycollab.core.utils.FileUtils
import com.mycollab.vaadin.resources.VaadinResource
import com.vaadin.server.DownloadStream
import com.vaadin.server.FileResource
import com.vaadin.server.Resource
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileInputStream
import java.io.IOException

/**
 * @author MyCollab Ltd
 * @since 6.0.0
 */
class VaadinFileResource : VaadinResource {
    companion object {
        val LOG = LoggerFactory.getLogger(VaadinFileResource::class.java)
    }


    override fun getStreamResource(documentPath: String): Resource = FileStreamDownloadResource(documentPath)

    class FileStreamDownloadResource(documentPath: String) : FileResource(File(FileUtils.homeFolder, documentPath)) {
        override fun getStream(): DownloadStream? {
            val fileName = filename.replace(" ", "_").replace("-", "_")
            try {
                val inStream = FileInputStream(sourceFile)
                val ds = DownloadStream(inStream, mimeType, fileName)
                ds.setParameter("Content-Disposition", "attachment; filename=" + fileName)
                ds.cacheTime = 0
                return ds
            } catch (e: IOException) {
                LOG.error("Error to create download stream", e)
                return null
            }
        }
    }
}